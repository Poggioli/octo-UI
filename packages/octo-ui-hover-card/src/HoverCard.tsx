import { forwardRef } from "@octo-ui/core";
import {
  StyledHoverCardArrow,
  StyledHoverCardContent,
  StyledHoverCardPortal,
  StyledHoverCardRoot,
  StyledHoverCardTrigger,
} from "./styles";
import {
  HoverCardArrowProps,
  HoverCardContentProps,
  HoverCardPortalProps,
  HoverCardProps,
  HoverCardTriggerProps,
} from "./types";

const HoverCard = forwardRef<typeof StyledHoverCardRoot, HoverCardProps>(
  ({ ...props }, forwardedRef) => {
    return <StyledHoverCardRoot {...props} ref={forwardedRef} />;
  }
);
HoverCard.toString = () => `.${StyledHoverCardRoot.className}`;

// ========================================================================= //

const HoverCardTrigger = forwardRef<
  typeof StyledHoverCardTrigger,
  HoverCardTriggerProps
>(({ ...props }, forwardedRef) => {
  return <StyledHoverCardTrigger {...props} ref={forwardedRef} asChild />;
});
HoverCardTrigger.toString = () => `.${StyledHoverCardTrigger.className}`;

// ========================================================================= //

const HoverCardPortal = forwardRef<
  typeof StyledHoverCardPortal,
  HoverCardPortalProps
>(({ ...props }, forwardedRef) => {
  return <StyledHoverCardPortal {...props} ref={forwardedRef} />;
});

// ========================================================================= //

const HoverCardArrow = forwardRef<
  typeof StyledHoverCardArrow,
  HoverCardArrowProps
>(({ ...props }, forwardedRef) => {
  return <StyledHoverCardArrow {...props} ref={forwardedRef} />;
});

// ========================================================================= //

const HoverCardContent = forwardRef<
  typeof StyledHoverCardContent,
  HoverCardContentProps
>(
  (
    {
      children,
      forceMountPortal,
      container,
      width,
      height,
      hasArrow = true,
      collisionPadding = 20,
      sideOffset = 5,
      ...props
    },
    forwardedRef
  ) => {
    return (
      <HoverCardPortal forceMount={forceMountPortal} container={container}>
        <StyledHoverCardContent
          {...props}
          ref={forwardedRef}
          collisionPadding={collisionPadding}
          sideOffset={sideOffset}
        >
          {children}
          {hasArrow ? (
            <HoverCardArrow width={width} height={height} data-testid="arrow" />
          ) : null}
        </StyledHoverCardContent>
      </HoverCardPortal>
    );
  }
);
HoverCardContent.toString = () => `.${StyledHoverCardContent.className}`;

export {
  StyledHoverCardArrow,
  StyledHoverCardContent,
  StyledHoverCardPortal,
  StyledHoverCardRoot,
  StyledHoverCardTrigger,
};
export { HoverCard, HoverCardTrigger, HoverCardContent };
