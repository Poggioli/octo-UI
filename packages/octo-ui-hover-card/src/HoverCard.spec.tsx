import { render } from "@testing-library/react";
import { HoverCard, HoverCardContent, HoverCardTrigger } from "./HoverCard";

describe("<HoverCard />", () => {
  it(`GIVEN a <HoverCard />
      WHEN render is called
      THEN should render a <HoverCard />`, () => {
    const rendered = render(
      <HoverCard defaultOpen>
        <HoverCardTrigger>
          <button>Trigger</button>
        </HoverCardTrigger>
        <HoverCardContent>
          <h3>HoverCard content</h3>
        </HoverCardContent>
      </HoverCard>
    );
    expect(rendered.container).toMatchSnapshot();
  });

  it(`GIVEN a <HoverCard />
      WHEN render is called
      THEN should render a <HoverCard />`, () => {
    const rendered = render(
      <HoverCard>
        <HoverCardTrigger>
          <button>Trigger</button>
        </HoverCardTrigger>
        <HoverCardContent>
          <h3>HoverCard content</h3>
        </HoverCardContent>
      </HoverCard>
    );
    expect(rendered.container).toMatchSnapshot();
  });

  describe("hasArrow", () => {
    it(`GIVEN a <HoverCard />
        WHEN hasArrow props is false
        THEN should NOT render a <HoverCardArrow />`, () => {
      const rendered = render(
        <HoverCard defaultOpen>
          <HoverCardTrigger>
            <button>Trigger</button>
          </HoverCardTrigger>
          <HoverCardContent hasArrow={false}>
            <h3>HoverCard content</h3>
          </HoverCardContent>
        </HoverCard>
      );
      const arrow = rendered.queryByTestId("arrow");
      expect(arrow).toBeNull();
    });

    it(`GIVEN a <HoverCard />
        WHEN hasArrow it's not setted
        THEN should render a <HoverCardArrow />`, () => {
      const rendered = render(
        <HoverCard defaultOpen>
          <HoverCardTrigger>
            <button>Trigger</button>
          </HoverCardTrigger>
          <HoverCardContent>
            <h3>HoverCard content</h3>
          </HoverCardContent>
        </HoverCard>
      );
      const arrow = rendered.queryByTestId("arrow");
      expect(arrow).toBeTruthy();
    });
  });

  it(`GIVEN a <HoverCard />
      WHEN toString is called
      THEN return a string`, () => {
    expect(HoverCard.toString()).toBe(".octo-UI-c-PJLV");
  });

  it(`GIVEN a <HoverCardContent />
      WHEN toString is called
      THEN return a string`, () => {
    expect(HoverCardContent.toString()).toBe(".octo-UI-c-dsEMLZ");
  });

  it(`GIVEN a <HoverCardTrigger />
      WHEN toString is called
      THEN return a string`, () => {
    expect(HoverCardTrigger.toString()).toBe(".octo-UI-c-PJLV");
  });
});
