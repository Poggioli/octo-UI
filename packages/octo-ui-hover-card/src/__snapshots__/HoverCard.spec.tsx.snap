// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HoverCard /> GIVEN a <HoverCard />
      WHEN render is called
      THEN should render a <HoverCard /> 1`] = `
<div>
  <button
    class="octo-UI-c-PJLV"
    data-state="open"
  >
    Trigger
  </button>
</div>
`;

exports[`<HoverCard /> GIVEN a <HoverCard />
      WHEN render is called
      THEN should render a <HoverCard /> 2`] = `
<div>
  <button
    class="octo-UI-c-PJLV"
    data-state="closed"
  >
    Trigger
  </button>
</div>
`;
